microsoft/PowerToys
# Add post-build event for publishing SettingsV2

## Summary of the Pull Request
This PR adds a post-build script to the Settings.UI.Runner project. The script runs a build of the project and its dependencies and publish with the required configuration as required for the current MSI. Comments have been added to the script with REM commands. Once this is confirmed to work, we can do the same for the launcher project as well.


## PR Checklist
This PR fixes multiple undocumented issues:
- The main build pipeline fails because of missing dependencies for the MSI.
- This possibly fixes the Xaml Island double compile issue (to be verified by more users)

## Detailed Description of the Pull Request / Additional comments
- Since dotnet publish CLI was always failing for the Settings.UI UWP project (according to the [official github repo](https://github.com/dotnet/cli), some functionality in VS steps is not present in dotnet CLI), I used MsBuild instead, because VS uses MsBuild internally
- For compiling correctly with MsBuild in the interop project a reference to `common` projects file had to be removed and was changed to a path based reference rather than project based reference. I am not sure why normally building with VS doesn't throw a warning or error about the same.
- There was a missing file in the UWP project which existed but was not referenced in the project.
- The script in short runs a MSBuild publish task (which internally does build followed by publish). Some extra code had to be added to ensure it doesnt run into a recurse loop, since the msbuild task is a post build event.
- For publishing, a publish profile file needs to be created, that has been hardcoded into the script.
- For changing the MSI from its current behaviour to self-contained we would have to change one parameter in the script, as well as add the extra dll files created into the Wix toolset file.
- Since the script code is only readable when opening it in VS here is the code

      rem Check required to avoid a recursive loop of building

      IF NOT DEFINED settingsBuildEventCompleted (
      setlocal enableDelayedExpansion
      SET settingsProfileFolderName=Properties\PublishProfiles\

      rem Create the publish profile folder if it doesn't exist
      IF NOT EXIST !settingsProfileFolderName! (mkdir !settingsProfileFolderName!)
      SET settingsProfileFileName=SettingsProfile.pubxml
      SET settingsPublishProfile=!settingsProfileFolderName!!settingsProfileFileName!

      rem Create the publish profile pubxml
      echo ^<?xml version="1.0" encoding="utf-8"?^> > !settingsPublishProfile!
      echo ^<^^!-- >> !settingsPublishProfile!
      echo https://go.microsoft.com/fwlink/?LinkID=208121.  >> !settingsPublishProfile!
      echo --^> >> !settingsPublishProfile!
      echo ^<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003"^> >> !settingsPublishProfile!
      echo   ^<PropertyGroup^> >> !settingsPublishProfile!
      echo     ^<PublishProtocol^>FileSystem^</PublishProtocol^> >> !settingsPublishProfile!
      echo     ^<Configuration^>$(ConfigurationName)^</Configuration^> >> !settingsPublishProfile!
      echo     ^<Platform^>x64^</Platform^> >> !settingsPublishProfile!
      echo     ^<TargetFramework^>netcoreapp3.1^</TargetFramework^> >> !settingsPublishProfile!
      echo     ^<PublishDir^>..\..\..\x64\$(ConfigurationName)\SettingsUIRunner^</PublishDir^> >> !settingsPublishProfile!
      echo     ^<RuntimeIdentifier^>win-x64^</RuntimeIdentifier^> >> !settingsPublishProfile!
      echo     ^<SelfContained^>false^</SelfContained^> >> !settingsPublishProfile!
      echo     ^<PublishSingleFile^>False^</PublishSingleFile^> >> !settingsPublishProfile!
      echo     ^<PublishReadyToRun^>False^</PublishReadyToRun^> >> !settingsPublishProfile!
      echo   ^</PropertyGroup^> >> !settingsPublishProfile!
      echo ^</Project^> >> !settingsPublishProfile!

      rem Set a variable to know that the publish event is called to avoid it being called again
      SET settingsBuildEventCompleted=1

      rem In case of Release we should not use Debug CRT in VCRT forwarders
      IF $(ConfigurationName)==Release (
      "$(MSBuildBinPath)\msbuild.exe" Microsoft.PowerToys.Settings.UI.Runner.csproj  -p:Configuration="$(ConfigurationName)" -p:Platform="$(PlatformName)" -p:AppxBundle=Never -p:VCRTForwarders-IncludeDebugCRT=false -p:PublishProfile=!settingsProfileFileName!
      ) ELSE (
      "$(MSBuildBinPath)\msbuild.exe" Microsoft.PowerToys.Settings.UI.Runner.csproj  -p:Configuration="$(ConfigurationName)" -p:Platform="$(PlatformName)" -p:AppxBundle=Never -p:PublishProfile=!settingsProfileFileName!
      )
      )


## Validation Steps Performed
- Build Debug and Release and tried building the MSI. Also ran PowerToys to verify settings opens. Also check the x64\Config\Settings.UI.Runner\*.deps.json file which should only contain win-x64 version of dlls.
